### DISCUSSION TITLE
builtin commands interactive with fzf

### DISCUSSION THREAD

**Discussion by jqhr** [nushell/nushell#10859] [Q&A] [answered]
For example.
```bash
 ls **/*.md | fzf
```
the result is
![图片](https://github.com/nushell/nushell/assets/23305067/dd9e1779-732a-46ab-84b9-fed9fc5a52d6)
How to fix it?

**Comment by amtoine** [comment]
what do you want to fix?
- the ANSI escape sequences: you can use `ansi strip` or maybe there is something like `--no-ansi` or `--ansi` on `fzf`
- the table formatting: you could do something like `ls ... | get name | fzf ...` or `ls ... | each { $"($in.name): ($in.modified)" } | fzf ...`

**Reply by jqhr** [reply]
Thanks, I think the input and output of fzf could prettier and be parsed. Only name or other field into fzf is not enough.
For example, I want find files both by name and modified by fzf.

**Comment by ghost** [comment, marked as answer]
fzf has no nushell integration. therefore nushell sends a nice looking table and fzf takes each line as input (including decodration, etc).  
you can run `ls | get name | str join "\n" | fzf`.  

i have added a custom command to my config, which is a nushell aware version of fzf:
```nu
def nufzf [] {$in | each {|i| $i | to json --raw} | str join "\n" | fzf  | from json}
```
it can be used like this: `ls | nufzf` or `ps | nufzf | kill $in.pid`.  
it uses `json` instead of `nuon` since `nuon` is really unrealiable and can't even handle `ls | to nuon --raw | from nuon`...

a explanation on how it works:
- `$in` reads the stdin (output of the prvious command in the pipeline)
- `| each {…}` iterates over each item in the output
- `$i | to json --raw` converts the items into `json` (a nushell parsable, human readable, fzf compatible format)
- ` | str join "\n"` converts the json lines from a list into one long fzf compatible string
- `| fzf` sends it to fzf
- `| from json` reads the selected item and converts it back from json to nushell data

**Reply by jqhr** [reply]
Great !

**Reply by sim590** [reply]
Selecting multiple values with FZF is often useful. I extended your definition to the following in order to support it:

```nushell
def nufzf [] {
  $in | each {|i| $i | to json --raw}
      | str join "\n"
      | fzf -m
      | lines
      | each {$in | from json}
      | reduce {|it, acc| $acc | append $it }
}
```

I'm a nushell novice, so I don't know if I could have made this simpler.

**Comment by jochumdev** [comment]
Updated this a bit:

```nu
# Fuzzy search any data with fzf in the pipe
# 
# Supports all container formats using json for complex types
# and normal lists for lists.
#
# Examples:
#
# ["foo", "bar"] | fzf
#
# Multiline:
# ["foo", "bar"] | fzf -m
##@complete external
def --wrapped fuzzy [
    ...args
]: list<any> -> list<any>, table -> table, record -> record, string -> string {
    let input = $in

    match ($input | describe | str replace --regex '<.*' '') {
        "list" => {
            $input | str join "\n" | ^fzf ...$args | lines
        },
        _ => {
            $input | each {|i| $i | to json --raw}
              | str join "\n"
              | ^fzf ...$args
              | lines
              | each {$in | from json}
              | reduce {|it, acc| $acc | append $it}
        }
    }
}
```

Will work future on it.
